File: 12_bermudan_swaption_lsm_pricing/cpp/RegressionBasis.cpp
#include "RegressionBasis.hpp"

#include <cmath>
#include <string>
#include <stdexcept>

/*
    Polynomial basis:

        1, x, x^2, ..., x^degree
*/

std::vector<double> polynomialBasis(
    double x,
    int degree
){

    std::vector<double> basis;

    basis.push_back(1.0);

    for(int d = 1; d <= degree; ++d){

        basis.push_back(
            std::pow(
                x,
                d
            )
        );
    }

    return basis;
}

/*
    Basis using both short rate and swap rate.

    degree = 2:

        1, r, s, r*s, r^2, s^2
*/

std::vector<double> rateAndSwapBasis(
    double shortRate,
    double swapRate,
    int degree
){

    std::vector<double> basis;

    basis.push_back(1.0);
    basis.push_back(shortRate);
    basis.push_back(swapRate);
    basis.push_back(shortRate * swapRate);

    if(degree >= 2){

        basis.push_back(shortRate * shortRate);
        basis.push_back(swapRate * swapRate);
    }

    if(degree >= 3){

        basis.push_back(shortRate * shortRate * shortRate);
        basis.push_back(swapRate * swapRate * swapRate);
        basis.push_back(shortRate * shortRate * swapRate);
        basis.push_back(shortRate * swapRate * swapRate);
    }

    return basis;
}

/*
    Build basis vector by basis type.
*/

std::vector<double> buildBasisVector(
    double shortRate,
    double swapRate,
    const std::string& basisType,
    int degree
){

    if(basisType == "rate"){

        return polynomialBasis(
            shortRate,
            degree
        );

    } else if(basisType == "swap"){

        return polynomialBasis(
            swapRate,
            degree
        );

    } else if(basisType == "rate_swap"){

        return rateAndSwapBasis(
            shortRate,
            swapRate,
            degree
        );

    } else {

        throw std::runtime_error(
            "basisType must be rate, swap, or rate_swap."
        );
    }
}

File: 12_bermudan_swaption_lsm_pricing/python/regression_basis.py
"""
regression_basis.py

Regression basis functions for Longstaff-Schwartz Bermudan swaption pricing.
"""

import numpy as np


def polynomial_basis(
    x,
    degree=2
):
    """
    Polynomial regression basis.

    Parameters
    ----------
    x : np.ndarray
        State variable, usually short rate or swap rate.

    degree : int
        Polynomial degree.

    Returns
    -------
    np.ndarray
        Design matrix.
    """

    x = np.asarray(x)

    columns = [
        np.ones_like(x)
    ]

    for d in range(1, degree + 1):

        columns.append(
            x ** d
        )

    return np.column_stack(
        columns
    )


def rate_and_swap_basis(
    short_rates,
    swap_rates,
    degree=2
):
    """
    Regression basis using both short rate and swap rate.

    This is useful because the continuation value of a Bermudan swaption
    depends on both the short-rate state and the moneyness of the swap.
    """

    r = np.asarray(short_rates)
    s = np.asarray(swap_rates)

    columns = [
        np.ones_like(r),
        r,
        s,
        r * s
    ]

    if degree >= 2:

        columns.extend(
            [
                r ** 2,
                s ** 2
            ]
        )

    if degree >= 3:

        columns.extend(
            [
                r ** 3,
                s ** 3,
                (r ** 2) * s,
                r * (s ** 2)
            ]
        )

    return np.column_stack(
        columns
    )